parachains/pallets/collective-content/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+		/// An on-chain announcement has been removed.
+		AnnouncementRemoved { cid: Cid },
+	}
+
+	/// The collective charter.
+	#[pallet::storage]
+	#[pallet::getter(fn charter)]
+	pub type Charter<T: Config<I>, I: 'static = ()> = StorageValue<_, Cid, OptionQuery>;
+
+	/// The collective announcements.
+	#[pallet::storage]
+	#[pallet::getter(fn announcements)]
+	pub type Announcements<T: Config<I>, I: 'static = ()> = StorageValue<
+		_,
+		BoundedVec<(Cid, Option<T::BlockNumber>), T::MaxAnnouncementsCount>,
+		ValueQuery,
+	>;
+
+	/// The closest expiration block number of an announcement.
+	#[pallet::storage]
+	pub type NextAnnouncementExpireAt<T: Config<I>, I: 'static = ()> =
+		StorageValue<_, T::BlockNumber, OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Set the collective  charter.
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::CharterOrigin].
+		/// - `cid`: [CID](super::Cid) of the IPFS document of the collective charter.
+		///
+		/// Weight: `O(1)`.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::set_charter())]
+		pub fn set_charter(origin: OriginFor<T>, cid: Cid) -> DispatchResult {
+			T::CharterOrigin::ensure_origin(origin)?;
+
+			Charter::<T, I>::put(&cid);
+
+			Self::deposit_event(Event::<T, I>::NewCharterSet { cid });
+			Ok(())
+		}
+
+		/// Publish an announcement.
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::CharterOrigin].
+		/// - `cid`: [CID](super::Cid) of the IPFS document to announce.
+		/// - `maybe_expire`: Expiration block of the announcement.
+		///
+		/// Weight: `O(1)`.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::announce(maybe_expire.map_or(0, |_| 1)))]
+		pub fn announce(
+			origin: OriginFor<T>,
+			cid: Cid,
+			maybe_expire: Option<DispatchTimeFor<T>>,
+		) -> DispatchResult {
+			T::AnnouncementOrigin::ensure_origin(origin)?;
+
+			let now = frame_system::Pallet::<T>::block_number();
+			let maybe_expire_at = maybe_expire.map(|e| e.evaluate(now));
+			ensure!(maybe_expire_at.map_or(true, |e| e > now), Error::<T, I>::InvalidExpiration);
+
+			let mut announcements = <Announcements<T, I>>::get();
+			announcements
+				.try_push((cid.clone(), maybe_expire_at.clone()))
+				.map_err(|_| Error::<T, I>::TooManyAnnouncements)?;
+			<Announcements<T, I>>::put(announcements);
+
+			if let Some(expire_at) = maybe_expire_at {
+				if NextAnnouncementExpireAt::<T, I>::get().map_or(true, |n| n > expire_at) {
+					NextAnnouncementExpireAt::<T, I>::put(expire_at);
+				}
+			}
+
+			Self::deposit_event(Event::<T, I>::AnnouncementAnnounced { cid, maybe_expire_at });
+			Ok(())
+		}
+
+		/// Remove an announcement.
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::CharterOrigin].
+		/// - `cid`: [CID](super::Cid) of the IPFS document to remove.
+		///
+		/// Weight: `O(1)`, less of the [Config::MaxAnnouncementsCount] is lower.
```

Comment:
worst case isn't it O(log(MaxAnnoucementsCount)) ?

Reply:
totally, binary search is ~logN complexity.

Reply:
Not, the `Weight` of this call is O(1).
You are talking about its complexity